Adds caching to notification api calls

ohadlevy opened the pull request:
this also removes pagination from the implementation, as that was not
implemented in the UI at all.

this implementation reduce the notification fetching time on
my production system from avg response times of 80ms to 20ms.

tbrisker commented on app/controllers/notification_recipients_controller.rb:
```diff
@@ -34,4 +26,27 @@ def find_resource
+    NotificationRecipient.
+      where(:user_id => User.current.id, :notification_id => Notification.active).
+      order(:created_at).
+      includes(:notification, :notification_blueprint).
```
You can use `.preload` or `.joins` to prevent a potentially large double left-outer-join

tbrisker commented on app/controllers/notification_recipients_controller.rb:
```diff
@@ -34,4 +26,27 @@ def find_resource
+
+  def load_notifications
+    NotificationRecipient.
+      where(:user_id => User.current.id, :notification_id => Notification.active).
```
Doesn't this cause an extra inner query on a table you (could be) joining in any case?

ohadlevy replied:
not sure i understand?

notification_id => Notification.scope adds conditions ? (expiry)

tbrisker replied:
I mean this is the current query:
```sql
SELECT "notification_recipients".* 
FROM "notification_recipients" 
WHERE "notification_recipients"."user_id" = $1 
AND "notification_recipients"."notification_id" IN (
    SELECT "notifications"."id" 
    FROM "notifications" 
    WHERE (notifications.expired_at >= '2017-03-20 14:57:26.205932'))  
ORDER BY "notification_recipients"."created_at" ASC
```
but if you anyways load the notifications, it makes sense to do a join and just add another condition and reduce the nested query.

tbrisker commented on app/models/notification_recipient.rb:
```diff
@@ -7,6 +7,7 @@ class NotificationRecipient < ActiveRecord::Base
   validates :user, :presence => true
 
   scope :unseen, -> { where(:seen => false) }
+  after_save :delete_user_cache
```
Is this the only condition that invalidates the cache?

ohadlevy replied:
thanks - you are correct, I need to invalidate the cache before the notification itself should expire - will fix.

tbrisker reviewed (commented):
:+1: in general, a few inline comments.

ohadlevy commented:
@tbrisker I've refactored the PR, Could you please re-review? thanks!

tbrisker commented on app/services/ui_notifications/cache_handler.rb:
```diff
@@ -0,0 +1,65 @@
+module UINotifications
+  class CacheHandler
+    def initialize(user_id = User.current)
```
`User.current.id`?

tbrisker replied:
missed this one

tbrisker commented on app/controllers/notification_recipients_controller.rb:
```diff
@@ -5,15 +5,8 @@ class NotificationRecipientsController < Api::V2::BaseController
-      :notifications => @notifications.paginate(paginate_options).map(&:payload),
-      :total => @notifications.count
-    }
+    payload = UINotifications::CacheHandler.new(User.current).payload
```
Shouldn't this be `User.current.id`?

tbrisker commented on app/services/ui_notifications/cache_handler.rb:
```diff
@@ -0,0 +1,65 @@
+        active.
+        joins(:notification_recipients).
+        where(notification_recipients: {user_id: user_id}).
+        minimum(:expired_at)
```
would it make sense to just expire the cache at the next expiring notification, regardless of the user, to simplify this lookup?

ohadlevy replied:
IMHO no, why are you concerned about this query?

tbrisker replied:
Might be premature optimization, but if the tables grow significantly this might become a painful join for every cache miss.

tbrisker commented on test/unit/ui_notifications/cache_handler_test.rb:
```diff
@@ -0,0 +1,34 @@
+    blueprint = FactoryGirl.create(:notification_blueprint, :expires_in => 3.hours,
+                              :message => 'this test just executed successfully')
+    notification = FactoryGirl.create(:notification, :notification_blueprint => blueprint, :audience => 'global')
+    FactoryGirl.create(:notification_recipient, :notification =>  notification, user_id: user.id)
```
NP: I'm not opposed to new-style hashes, but at least be consistent with one style per line :smile:

tbrisker commented on test/unit/ui_notifications/cache_handler_test.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  test 'should calculate cache expiry time based on the notification expiry' do
+    user = FactoryGirl.create(:user)
```
Any reason to create a new user here rather then just using an existing one?

tbrisker commented on test/unit/ui_notifications/cache_handler_test.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  test 'should use cache if it exists' do
+  end
```
So does it use cache if it exists?

tbrisker reviewed (changes requested):
Looks better, a few inline comments.

ohadlevy commented:
@tbrisker I think I've addressed all of your comments..
![bart-simpson-chalkboard-wallpaper-generator](https://cloud.githubusercontent.com/assets/20167/24355468/af86fc00-12fe-11e7-82b6-7c7a949fcf0c.png)

ohadlevy commented:
@tbrisker updated again, also added .freeze on cache key and try in case User.current is nil.

ohadlevy commented on app/services/ui_notifications/cache_handler.rb:
```diff
@@ -0,0 +1,65 @@
+
+    def notifications
+      @notifications ||= NotificationRecipient.
+        where(:user_id => user_id, :notification_id => Notification.active).
```
I'll update the hash rocket syntax after tests will pass/fail.

ohadlevy committed:
fixes #18940 - adds caching to notification api calls.

this implementation reduce the notification fetching time on
my production system from avg response times of 80ms to 20ms,
and siginficantly reducing server load (db, cpu etc)

it also removes total counts from the implementation, as that was not
implemented in the UI at all.

ohadlevy commented:
@tbrisker test failures seems unrelated.

tbrisker commented:
Thanks @ohadlevy !
